server/src/db/queries/pathway/modifyPathwaySteps.js

```diff
@@ -0,0 +1,119 @@
+const Query = require('../../core/Query')
+const Model = require('../../core/Model')
+
+const convertToProp = (step) => {
+    let prop = {
+        ...step,
+        stepDataId: Model.generateId('StepData')
+    }
+
+    if (!step.isPathway) {
+        prop.stepContentId = Model.generateId('StepContent')
+    } 
+
+    delete step.isPathway
+    return prop
+}
+
+const createPathwaySteps = ({id, createSteps}) => {
+    const props = {
+        pathwayProps: [],
+        contentProps: []
+    }
+
+    createSteps.forEach((step) => {
+        step.isPathway ? props.pathwayProps.push(convertToProp(step))
+                       : props.contentProps.push(convertToProp(step))
+    })
+
+    console.log("Creating new steps ran")
+    console.log(props)
+
+    return new Query({
+        statement: `
+        MATCH (p:Pathway {id: $id})
+            FOREACH (prop IN $contentProps | 
+                CREATE (sd:Step {id: prop.stepDataId, name: prop.name, difficulty: prop.difficulty, index: prop.index, isPathway: false})
+                CREATE (sc:Content {id: prop.stepContentId, content: prop.content})
+                MERGE (p)<-[:HAS_PARENT_PATHWAY]-(sd)-[:HAS_CONTENT]->(sc)
+            )
+        WITH p
+            FOREACH (prop IN $pathwayProps | 
+                CREATE (sd:Step {id: prop.stepDataId, name: prop.name, difficulty: prop.difficulty, index: prop.index, isPathway: true})
+                MERGE (p1:Pathway {id: prop.pathwayId})
+                MERGE (p)<-[:HAS_PARENT_PATHWAY]-(sd)-[:INCLUDES]->(p1)
+            )
+        RETURN p
+        `,
+        params: {
+            pathwayProps: props.pathwayProps,
+            contentProps: props.contentProps,
+            id,
+        }
+    })
+}
+
+const updatePathwaySteps = ({id, updateSteps}) => {
+    const steps = []
+    updateSteps.forEach((step) => {
+        steps.push(convertToProp(step));
+    })
+
+    console.log("Updating steps")
+    console.log(steps)
+
+    return new Query({
+        statement: `
+        MATCH (p:Pathway {id: $id})
+        WITH p
+        UNWIND $steps as upd
+        WITH p, upd
+        MATCH (p)<-[:HAS_PARENT_PATHWAY]-(s:Step {id: upd.id})
+        SET s.name = upd.name, s.difficulty = upd.difficulty, s.index = upd.index
+        RETURN p
+        `,
+        params: {
+            steps,
+            id,
+        }
+    })
+
+}
+
+const deletePathwaySteps = ({id, deleteSteps}) => {
+    const steps = []
+    deleteSteps.forEach((step) => {
+        steps.push(convertToProp(step));
+    })
+
+    console.log("Deleting steps")
+    console.log(steps)
+
+    return new Query({
+        statement: `
+        MATCH (p:Pathway {id: $id})
+        WITH p
+        UNWIND $steps as del
+        WITH p, del
+        MATCH (p)<-[:HAS_PARENT_PATHWAY]-(s:Step{id: del.id})
```

Comment:
This query will delete the node that contains the metadata, but in case that node does not point to a pathway, it will not delete the node containing the content. Not desired. You should delete the content node too if the Step node is not pointing to a pathway.

Reply:
Again going by the three layer style, the actual content of the step has only one instance i.e. an md file. This step is just a pointer to it, deleting the pointer should suffice.